### PR TITLE
Spawn-on-demand implementation

### PR DESCRIPTION
**Pull request by hjoliver**
These changes partially address #3304 as specified in the proposal https://github.com/cylc/cylc-admin/pull/70

Replaces the old spawn-on-submit (SOS) method of evolving the workflow with spawn-on-demand (SOD): when task output messages are received, downstream dependent tasks are spawned (if neccessary) and their prerequisites immediately updated. (See the proposal for the historical basis of the current system).

Advantages include:
- dramatic reduction in task pool size (and therefore also in the number of tasks that the UI sees) 
  -  (no tasks waiting before they are needed, and succeeded tasks can be removed immediately)
  - OLD: at least one instance of every task in the workflow
  - NEW: only current active tasks
  - (e.g. a suite with 10k tasks per cycle might have only 10s or 100s active at once!)
- no need for dependency matching
- no need for task proxy life-cycle management (clean-up cut-off computation)
- graph branching: no need for suicide triggers
- tasks can run out of order if needed (so no more stalls due to unspawned successors downstream of a failed task)

**TODO:** 
- **redesign the internals specifically for SOD**
  - (this is a quick implementation using and abusing the existing machinery)
  - new task pool structure, to avoid iteration?
    - e.g. dict keyed by task ID?
  - direct update of prerequisites without iteration?
  - do we still need separate prerequisite and output classes?
- **extensive testing** and **consideration of edge cases and consequences**
  - (Tested so far with some basic suites; seems to run the complex demo fine; have not looked at the test battery but expect a lot of failures due to e.g. different task pool and DB content)
- check how this works with the new UI; from a quick test @dwsutherland's extra nodes (n=1 rather than n=0? at the back of the workflow?) add a lot of succeeded tasks to the pool that the UI sees ... so we might need to revisit that. 

(I won't take this further until the new internals design is worked through and agreed)

**Consequences for users**

If this works out, the major consequence for users (apart from better performance!) will be no more (almost) waiting or succeeded tasks visible in the UI, because they will no longer be present in the task pool. However, note that:
- the presence or absence of waiting or succeeded tasks in a particular cycle point, currently, is essentially non-explicable to users (we do not populate whole cycle points!), but the new n=0 window makes perfect sense: it's the active tasks
- we just need to allow users to see what comes next, or what came before, in the workflow by more sensible means:
  - expand from n=0 to n=1,2,3  window (by graph edges)
  - click on a task and see the graph connections around it

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).

- [ ] Appropriate tests are included (unit and/or functional).
- [ ] Already covered by existing tests.

- [ ] Appropriate change log entry included.

- [ ] (master branch) I have opened a documentation PR at cylc/cylc-doc/pull/XXXX.

### CODE REVIEW

**Comment by hjoliver** [comment]
(Provocatively assigning to 8.0.0 milestone :grin:)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- cylc/flow/task_events_mgr.py  8
         

Complexity decreasing per file
==============================
+ cylc/flow/dump.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4772278&bid=15902513)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- cylc/flow/task_events_mgr.py  8
         

Complexity decreasing per file
==============================
+ cylc/flow/dump.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4772278&bid=15902513)

**Comment by hjoliver** [comment]
## Graph branching example

```
[scheduling]
   cycling mode = integer
   initial cycle point = 1
   final cycle point = 2
   [[graph]]
       P1 = """amodel:failed => bad
               amodel => good
               bad | good => end"""
[runtime]
   [[root]]
       pre-script = "sleep 5"
   [[amodel]]
       script = "if (( CYLC_TASK_CYCLE_POINT == 2 )
```
sod-sui

**Comment by hjoliver** [comment]
## a non-trivial test suite

Multiple integer cycles, a parameterized cycle within each integer cycle, graph isolates, and tasks with no prerequisites.
```
[cylc]
   [[parameters]]
       i = 1..10
[scheduling]
   cycling mode = integer
   initial cycle point = 1
   final cycle point = 5
   [[graph]]
       R1/^/P1 = "start => afoo"
       R1/5/P1 = "zzz"
       P1 = """get => afoo => bar & baz
          afoo[-P1] => afoo
          baz:hello => qux<i=1>
          qux<i-1> => qux<i>
          qux<i=5> => flux"""
       P2 = "flux => xxx"
[runtime]
   [[root]]
       script = sleep 10
   [[qux<i>]]
       script = sleep 2
   [[baz]]
       script = """sleep 5
          cylc message "hello world"
          sleep 5"""
       [[[outputs]]]
          hello = "hello world"
```
At run time: top panel spawn-on-demand, bottom panel spawn-on-submit (at a similar point in the workflow).
shot

**Comment by hjoliver** [comment]
## the "complex" demo suite

spawn-on-demand (managing ~100 tasks)
shot-complex-sod

spawn-on-submit (managing ~1500 tasks)
shot-complex-sos

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- cylc/flow/task_events_mgr.py  8
         

Complexity decreasing per file
==============================
+ cylc/flow/dump.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4772278&bid=15902513)

**Comment by kinow** [comment]
Quick memory profile shows that this branch is using less memory than `master` in Cylc 8 running the complex suite (no tweaks).

Cylc 8 `master`:

master

This branch:

Figure_4

The suite got stalled after ~10 minutes, but I didn't look at logs or anything as this is still WIP.

The suite `five` works with no issues, but there is little difference. One interesting thing was that one of the jobs appears to have reached a memory peak higher than on `master`, though that could be some local issue.

`master` running `five`.

Figure_2

This branch running `five`:

Figure_3

Used `memory_profiler` module to measure, with the following command syntax:

```
mprof run --include-children cylc run --no-detach complex
```

I've modified the `five` locally to run over a single day. And the command above needs `include-children`, otherwise you get the memory profile of only the scheduler, not the jobs.

:+1: looking good so far!

**Comment by dwsutherland** [comment]
Well done!

> from a quick test @dwsutherland's extra nodes (n=1 rather than n=0? at the back of the workflow?) add a lot of succeeded tasks to the pool that the UI sees ... so we might need to revisit that.
> we just need to allow users to see what comes next, or what came before, in the workflow by more sensible means:
expand from n=0 to n=1,2,3 window (by graph edges)
click on a task and see the graph connections around it

The data store will generate all nodes/edges in for each existing cycle point in the pool , so I suspect if you want to know what's coming you may need to set `ghosts` to `true` in the GraphQL args.. However, the future ghosts won't have information like `prerequisites`/`conditions` (and their status) as that's not collected until the task is in the pool, although the information is there in part through the edges..

The deletion from the data-store is also cycle wise, so once a cycle has no tasks, the respective nodes/edges are deleted.

@hjoliver - What does holding a task mean now? Don't spawn the next task(s) that depend on the held one(s)?

**Comment by hjoliver** [comment]
@kinow's profiling results are interesting, I'll need to look closer, but a few comments and speculations: 

- The larger the suite, and the "longer" the graph within a cycle, and the more spread out it is over cycle points, the bigger the difference this change will make.  It won't make much difference for a small suite like "five" with little spread over cycle points.

- It's not surprising that the initial memory spike at start-up remains, as we still have to parse the suite generate the same number of "task definitions". At run-time, however, the number of task proxy objects instantiated will be greatly reduced, and there should only be spikes associated with large numbers of tasks coming "into demand" at once. So the big spike to 800 MB after 300 secs is surprising...

@dwsutherland:

> The data store will generate all nodes/edges for each existing cycle point in the pool 

... so I suspect this (the new data store for the new UI) is artificially spiking memory use by suddenly "fleshing out" an entire cycle point as soon as a task hits the next cycle point.

**Comment by hjoliver** [comment]
> And the command above needs include-children, otherwise you get the memory profile of only the scheduler, not the jobs.

I think we only want to profile the scheduler here, not the jobs.  This change should not make any difference to the evolution of the workflow (except in the rare circumstance that the old way stalls due to un-spawned successors downstream of a failed task) and no change to the jobs.  It should just greatly reduce the size of the pool of task proxies that the scheduler has to manage.

**Comment by hjoliver** [comment]
> The suite got stalled after ~10 minutes, 

I saw this too.  It may be to the one major problem with spawn-on-demand that I have identified by not fixed yet (which actually makes my statement above untrue for the current state of the branch: "This change should not make any difference to the evolution of the workflow"): any conditional dependencies (including, for e.g. `FAMILY:succeed-any`) will likely cause strange behaviour).

**Comment by hjoliver** [comment]
@dwsutherland 

> @hjoliver - What does holding a task mean now? Don't spawn the next task(s) that depends any held one(s)?

Hold just means don't run (i.e. don't submit your job) even if you are ready to.  And if you don't run you won't generate any outputs, and therefore won't spawn any downstream tasks.

**Comment by hjoliver** [comment]
> @dwsutherland:
> 
> > The data store will generate all nodes/edges for each existing cycle point in the pool
> 
> ... so I suspect this (the new data store for the new UI) is artificially spiking memory use by
> suddenly "fleshing out" an entire cycle point as soon as a task hits the next cycle point.

Actually, my tests suggest that this doesn't have a much effect (phew); so the spikes are probably caused by @kinow 's profiling of local server-submitted jobs as well as the server itself.

**Comment by hjoliver** [comment]
Here's my quick test, complex suite with default queue limit of 10 ('cos they're local tasks), new data-store update commented out, and profiling the server only:

## master
Figure_3

## this branch
Figure_1

This suggests ~25% memory saving on the complex suite - **at a minimum** because the test has  very light weight task proxies (all the task configuration has been stripped out of the complex suite) and few active cycle points.  That sounds kind of reasonable, or at least not unexpected ... that the task pool could take up about that much space for this kind of suite.

**Comment by hjoliver** [comment]
BTW, server memory savings are great, but not the primary reason to do spawn-on-demand. The main things are:
- a better (and much easier to understand) model of workflow evolution
- much reduced (and much easier to understand) number of nodes/tasks for the UI to display

**Comment by kinow** [comment]
>     * much reduced (and much easier to understand) number of nodes/tasks for the UI to display

Yup, and "much reduced (and much easier to understand) number of nodes/tasks for the UI to display" will be a great help for the UI development - this doesn't mean we won't try to optimize as much as possible, just that this will make the overall experience to users even better :)

**Comment by hjoliver** [comment]
To illustrate task pool bloat with number of active cycles:

```
[cylc]
   [[parameters]]
       m = 0..20
       n = 0..20
[scheduling]
   initial cycle point = 3000
   max active cycle points = 10  # <----
   [[queues]]
      [[[default]]]
         limit = 20
   [[graph]]
      P1Y = """foo[-P1Y] => foo => bar<m> => qux<m=1,n> => waz
            qux<m-1,n> => qux<m,n>"""
```
Tasks per cycle = 464

## master 
t2-master

## this branch
t2-sod

i.e., with 10 active cycles, master -> 120 MB; this branch -> 60 MB; the difference is purely task pool size

**Comment by hjoliver** [comment]
(Both still rising however ... I wonder if we have a leak somewhere?)

**Comment by kinow** [comment]
> (Both still rising however ... I wonder if we have a leak somewhere?)

Could it be the jobs in the data store?

After a `pip install objgraph` in my venv, and this diff applied to `scheduler.py`:

```diff
diff --git a/cylc/flow/scheduler.py b/cylc/flow/scheduler.py
index 9e02121ea..7ce5042ef 100644
--- a/cylc/flow/scheduler.py
+++ b/cylc/flow/scheduler.py
@@ -98,6 +98,33 @@ from cylc.flow.wallclock import (
     get_utc_mode)
 from cylc.flow.xtrigger_mgr import XtriggerManager
 
+import tracemalloc
+import objgraph
+
+
+def display_top(snapshot, group_by='lineno', limit=10):
+    snapshot = snapshot.filter_traces((
+        tracemalloc.Filter(False, "<frozen importlib._bootstrap>"),
+        tracemalloc.Filter(False, "<unknown>"),
+    ))
+    top_stats = snapshot.statistics(group_by)
+
+    print("Top %s lines" % limit)
+    for index, stat in enumerate(top_stats[:limit], 1):
+        frame = stat.traceback[0]
+        # replace "/path/to/module/file.py" with "module/file.py"
+        filename = os.sep.join(frame.filename.split(os.sep)[-2:])
+        print("#%s: %s:%s: %.1f KiB"
+              % (index, filename, frame.lineno,
+                 stat.size / 1024))
+
+    other = top_stats[limit:]
+    if other:
+        size = sum(stat.size for stat in other)
+        print("%s other: %.1f KiB" % (len(other), size / 1024))
+    total = sum(stat.size for stat in top_stats)
+    print("Total allocated size: %.1f KiB" % (total / 1024))
+
 
 class SchedulerStop(CylcError):
     """Scheduler normal stop."""
@@ -1566,6 +1593,7 @@ see `COPYING' in the Cylc source distribution.
         self.initialise_scheduler()
         self.ws_data_mgr.initiate_data_model()
         self.publisher.publish(self.ws_data_mgr.get_publish_deltas())
+        tracemalloc.start()
         while True:  # MAIN LOOP
             tinit = time()
             has_reloaded = False
@@ -1632,7 +1660,11 @@ see `COPYING' in the Cylc source distribution.
             elif quick_mode:
                 sleep(self.INTERVAL_MAIN_LOOP_QUICK - elapsed)
             else:
-                sleep(self.INTERVAL_MAIN_LOOP - elapsed)
+                objgraph.show_most_common_types(limit=12)
+                objgraph.show_growth(limit=12)
+                snapshot = tracemalloc.take_snapshot()
+                display_top(snapshot)
+                sleep(self.INTERVAL_MAIN_LOOP * 10 - elapsed)  # sleep longer
             # Record latest main loop interval
             self.main_loop_intervals.append(time() - tinit)
             # END MAIN LOOP
```

Prints the following on the first main loop longer sleep:

```bash
list                    15355
function                14196
dict                    11917
tuple                   6381
OrderedDictWithDefaults 6097
set                     3363
weakref                 2794
wrapper_descriptor      2358
getset_descriptor       2133
FuncCodeInfo            1863
type                    1690
method_descriptor       1632
list                       15355    +15355
function                   14196    +14196
dict                       11792    +11792
tuple                       6337     +6337
OrderedDictWithDefaults     6097     +6097
set                         3368     +3368
weakref                     2794     +2794
wrapper_descriptor          2358     +2358
getset_descriptor           2133     +2133
FuncCodeInfo                1868     +1868
type                        1690     +1690
method_descriptor           1632     +1632
Top 10 lines
#1: flow/data_store_mgr.py:139: 2679.6 KiB
#2: flow/task_id.py:46: 246.7 KiB
#3: flow/data_store_mgr.py:439: 162.5 KiB
#4: flow/task_proxy.py:317: 130.5 KiB
#5: flow/task_proxy.py:222: 120.4 KiB
#6: flow/data_store_mgr.py:659: 106.7 KiB
#7: flow/data_store_mgr.py:653: 102.0 KiB
#8: flow/task_outputs.py:72: 100.3 KiB
#9: flow/task_outputs.py:73: 61.9 KiB
#10: cycling/__init__.py:104: 57.4 KiB
619 other: 1274.0 KiB
Total allocated size: 5042.0 KiB
```

Then the following in the next run:

```bash
list                       16286
function                   14195
dict                       12267
tuple                      7163
OrderedDictWithDefaults    6097
set                        3429
weakref                    2810
wrapper_descriptor         2358
getset_descriptor          2133
FuncCodeInfo               1906
type                       1690
builtin_function_or_method 1639
list                          16286      +931
Prerequisite                    440      +352
tuple                          6518      +181
dict                          11895      +103
set                            3429       +61
FuncCodeInfo                   1906       +38
ISO8601Point                    276       +24
builtin_function_or_method     1639       +23
TaskProxy                       253       +21
TaskState                       253       +21
TaskOutputs                     253       +21
weakref                        2810       +16
Top 10 lines
#1: flow/data_store_mgr.py:139: 2932.8 KiB
#2: flow/task_id.py:46: 274.1 KiB
#3: flow/data_store_mgr.py:439: 180.5 KiB
#4: flow/task_proxy.py:317: 142.3 KiB
#5: flow/task_proxy.py:222: 141.2 KiB
#6: flow/task_outputs.py:72: 120.7 KiB
#7: flow/task_outputs.py:73: 72.6 KiB
#8: cycling/__init__.py:104: 61.1 KiB
#9: flow/task_pool.py:1012: 58.8 KiB
#10: cycling/loader.py:69: 56.4 KiB
811 other: 1424.2 KiB
Total allocated size: 5464.6 KiB
```

(function copied from [here](https://tech.buzzfeed.com/finding-and-fixing-memory-leaks-in-python-413ce4266e7d?gi=dfdb32c0d78b))

The main offender in the heap growth was on `flow/data_store_mgr.py:139`, going from 2679.6 KiB to 2932.8 KiB. Adding a line like `print(f"Data Store jobs: {len(self.ws_data_mgr.data['kinow|hmem1']['jobs'])}")` in the main loop, it shows that the length of jobs grows 20 after a few executions, then 20 again, and so it goes.

After a few minutes it went up to 200.

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- cylc/flow/task_events_mgr.py  8
         

Complexity decreasing per file
==============================
+ cylc/flow/dump.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4772278&bid=15902513)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- cylc/flow/task_events_mgr.py  8
         

Complexity decreasing per file
==============================
+ cylc/flow/dump.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4772278&bid=15902513)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- cylc/flow/task_events_mgr.py  8
         

Complexity decreasing per file
==============================
+ cylc/flow/dump.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4772278&bid=15902513)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- cylc/flow/task_events_mgr.py  8
         

Complexity decreasing per file
==============================
+ cylc/flow/dump.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4772278&bid=15902513)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- cylc/flow/task_events_mgr.py  8
         

Complexity decreasing per file
==============================
+ cylc/flow/dump.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4772278&bid=15902513)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- cylc/flow/task_events_mgr.py  8
         

Complexity decreasing per file
==============================
+ cylc/flow/dump.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4772278&bid=15902513)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- cylc/flow/task_events_mgr.py  8
         

Complexity decreasing per file
==============================
+ cylc/flow/dump.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4772278&bid=15902513)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- cylc/flow/task_events_mgr.py  6
         

Complexity decreasing per file
==============================
+ cylc/flow/taskdef.py  -6
+ cylc/flow/dump.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4772278&bid=15902513)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- cylc/flow/task_events_mgr.py  5
         

Complexity decreasing per file
==============================
+ cylc/flow/dump.py  -1
+ cylc/flow/taskdef.py  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4772278&bid=15902513)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- cylc/flow/task_events_mgr.py  5
         

Complexity decreasing per file
==============================
+ cylc/flow/dump.py  -1
+ cylc/flow/taskdef.py  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4772278&bid=15902513)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- cylc/flow/task_events_mgr.py  5
         

Complexity decreasing per file
==============================
+ cylc/flow/taskdef.py  -6
+ cylc/flow/dump.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4772278&bid=15902513)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- cylc/flow/task_events_mgr.py  5
         

Complexity decreasing per file
==============================
+ cylc/flow/dump.py  -1
+ cylc/flow/taskdef.py  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4772278&bid=15902513)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- cylc/flow/task_events_mgr.py  5
         

Complexity decreasing per file
==============================
+ cylc/flow/dump.py  -1
+ cylc/flow/taskdef.py  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4772278&bid=15902513)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- cylc/flow/task_events_mgr.py  5
         

Complexity decreasing per file
==============================
+ cylc/flow/taskdef.py  -6
+ cylc/flow/dump.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4772278&bid=15902513)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- cylc/flow/task_events_mgr.py  5
         

Complexity decreasing per file
==============================
+ cylc/flow/dump.py  -1
+ cylc/flow/taskdef.py  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4772278&bid=15902513)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 6
           

Complexity increasing per file
==============================
- cylc/flow/task_events_mgr.py  5
         

Complexity decreasing per file
==============================
+ cylc/flow/taskdef.py  -6
+ cylc/flow/dump.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4772278&bid=15902513)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 6
           

Complexity increasing per file
==============================
- cylc/flow/task_events_mgr.py  5
- cylc/flow/scripts/cylc_spawn.py  2
         

Complexity decreasing per file
==============================
+ cylc/flow/taskdef.py  -6
+ cylc/flow/dump.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4772278&bid=15902513)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 6
           

Complexity increasing per file
==============================
- cylc/flow/scripts/cylc_spawn.py  2
- cylc/flow/task_events_mgr.py  5
         

Complexity decreasing per file
==============================
+ cylc/flow/taskdef.py  -6
+ cylc/flow/dump.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4772278&bid=15902513)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 6
           

Complexity increasing per file
==============================
- cylc/flow/scripts/cylc_spawn.py  2
- cylc/flow/task_events_mgr.py  5
         

Complexity decreasing per file
==============================
+ cylc/flow/taskdef.py  -6
+ cylc/flow/dump.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4772278&bid=15902513)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 6
           

Complexity increasing per file
==============================
- cylc/flow/scripts/cylc_spawn.py  2
- cylc/flow/task_pool.py  3
- cylc/flow/task_events_mgr.py  5
         

Complexity decreasing per file
==============================
+ cylc/flow/dump.py  -1
+ cylc/flow/taskdef.py  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4772278&bid=15902513)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 6
           

Complexity increasing per file
==============================
- cylc/flow/scripts/cylc_spawn.py  2
- cylc/flow/task_events_mgr.py  5
- cylc/flow/task_pool.py  6
         

Complexity decreasing per file
==============================
+ cylc/flow/dump.py  -1
+ cylc/flow/taskdef.py  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4772278&bid=15902513)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 6
           

Complexity increasing per file
==============================
- cylc/flow/task_events_mgr.py  5
- cylc/flow/task_pool.py  6
- cylc/flow/scripts/cylc_spawn.py  2
         

Complexity decreasing per file
==============================
+ cylc/flow/taskdef.py  -6
+ cylc/flow/dump.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4772278&bid=15902513)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           

Complexity increasing per file
==============================
- cylc/flow/task_proxy.py  4
- cylc/flow/task_events_mgr.py  5
- cylc/flow/scripts/cylc_spawn.py  2
- cylc/flow/task_pool.py  6
         

Complexity decreasing per file
==============================
+ cylc/flow/dump.py  -1
+ cylc/flow/taskdef.py  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4772278&bid=15902513)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           

Complexity increasing per file
==============================
- cylc/flow/task_pool.py  6
- cylc/flow/task_proxy.py  4
- cylc/flow/scripts/cylc_spawn.py  2
- cylc/flow/task_events_mgr.py  5
         

Complexity decreasing per file
==============================
+ cylc/flow/dump.py  -1
+ cylc/flow/taskdef.py  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4772278&bid=15902513)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 7
           

Complexity increasing per file
==============================
- cylc/flow/task_proxy.py  4
- cylc/flow/task_pool.py  6
- cylc/flow/scripts/cylc_spawn.py  2
- cylc/flow/task_events_mgr.py  10
         

Complexity decreasing per file
==============================
+ cylc/flow/taskdef.py  -6
+ cylc/flow/dump.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4772278&bid=15902513)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 7
           

Complexity increasing per file
==============================
- cylc/flow/scripts/cylc_spawn.py  2
- cylc/flow/task_proxy.py  6
- cylc/flow/task_events_mgr.py  10
         

Complexity decreasing per file
==============================
+ cylc/flow/config.py  -2
+ cylc/flow/taskdef.py  -6
+ cylc/flow/dump.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4772278&bid=15902513)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 10
           

Complexity increasing per file
==============================
- cylc/flow/task_events_mgr.py  1
- cylc/flow/task_proxy.py  5
- cylc/flow/scripts/cylc_spawn.py  2
         

Complexity decreasing per file
==============================
+ cylc/flow/taskdef.py  -6
+ cylc/flow/dump.py  -1
+ cylc/flow/config.py  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4772278&bid=15902513)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 10
           

Complexity increasing per file
==============================
- cylc/flow/task_events_mgr.py  1
- cylc/flow/task_proxy.py  5
- cylc/flow/scripts/cylc_spawn.py  2
         

Complexity decreasing per file
==============================
+ cylc/flow/taskdef.py  -6
+ cylc/flow/dump.py  -1
+ cylc/flow/config.py  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4772278&bid=15902513)
